### PR TITLE
[STACK-2618] refine the controller_worker.py db_exceptions.NoResultFound retry 

### PR DESCRIPTION
**Pull request by sana-a10**
## Description
Use **tenacity.retry()** mechanism to wait for DB entry instead of **wait_for_db_entry()**

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2618

## Technical Approach
- Reverted the code added for **wait_for_db_entry()** api and used **tenacity.retry()** to wait for the Database entry
- Added following options required for tenacity.retry() in config_options.py file 
             - retry_attempts, 
             - retry_initial_delay, 
             - retry_bakcoff, 
             - retry_max

## Config Changes
<pre>
[a10_controller_worker]
amp_image_owner_id = e1fe759747844dd1871092efe9079a26
amp_secgroup_list = lb-mgmt-sec-grp
amp_flavor_id = d08431af-865b-4c38-8cee-7877f194cdb0
amp_boot_network_list = 01351f15-3adf-4183-a352-e30f29663007
amp_ssh_key_name = octavia_ssh_key
network_driver = a10_octavia_neutron_driver
workers = 2
amp_active_retries = 20
amp_active_wait_sec = 10
amp_image_id = 0788fe86-39af-4814-89ea-b444ae5f19a9
amp_image_tag = vthunder_5_2_1
loadbalancer_topology = SINGLE
<b>
db_retry_attempts = 20
db_retry_initial_delay = 2
db_retry_delay_backoff = 2
db_retry_max_delay = 6
</b>
</pre>

## Test Cases
- Create a LB, Listener, Pool, Member, Health Monitor, L7rule and L7policy

## Manual Testing

Create a Load balancer

```
stack@sana:~$ openstack loadbalancer create --vip-subnet-id vip_subnet --name lb1

+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| admin_state_up      | True                                 |
| availability_zone   | None                                 |
| created_at          | 2021-07-09T12:36:35                  |
| description         |                                      |
| flavor_id           | None                                 |
| id                  | ed1c3237-4a70-435d-88a1-7ad2039fbe8b |
| listeners           |                                      |
| name                | lb1                                  |
| operating_status    | OFFLINE                              |
| pools               |                                      |
| project_id          | e1fe759747844dd1871092efe9079a26     |
| provider            | a10                                  |
| provisioning_status | PENDING_CREATE                       |
| updated_at          | None                                 |
| vip_address         | 192.168.12.146                       |
| vip_network_id      | 144533bf-db26-440f-a0ca-8ba072769e4a |
| vip_port_id         | aa35d81c-eff6-4346-9424-0e842f02d34d |
| vip_qos_policy_id   | None                                 |
| vip_subnet_id       | 8330780e-25fd-4ec7-a6dd-fd412ec81e48 |
+---------------------+--------------------------------------+

stack@sana:~$

```
```
stack@sana:~$ openstack server list

+--------------------------------------+----------------------------------------------+--------+-----------------------------------------------------+----------------+-----------------+
| ID                                   | Name                                         | Status | Networks                                            | Image          | Flavor          |
+--------------------------------------+----------------------------------------------+--------+-----------------------------------------------------+----------------+-----------------+
| b20e8fee-8493-4863-9202-1caa358fb1a1 | amphora-305baec4-91f3-4d79-8b15-8c852c1f905e | ACTIVE | lb-mgmt-net=192.168.0.11; vip_network=192.168.12.12 | vThunder.qcow2 | vThunder_flavor |
+--------------------------------------+----------------------------------------------+--------+-----------------------------------------------------+----------------+-----------------+
stack@sana:~$

```
**Create a Listener**

```
stack@sana:~$ openstack loadbalancer listener create --protocol HTTP --protocol-port 91 lb1 --name l1

+-----------------------------+--------------------------------------+
| Field                       | Value                                |
+-----------------------------+--------------------------------------+
| admin_state_up              | True                                 |
| connection_limit            | -1                                   |
| created_at                  | 2021-07-09T12:46:46                  |
| default_pool_id             | None                                 |
| default_tls_container_ref   | None                                 |
| description                 |                                      |
| id                          | 42199b41-9abe-4554-b9bf-cc3fc6a96432 |
| insert_headers              | None                                 |
| l7policies                  |                                      |
| loadbalancers               | ed1c3237-4a70-435d-88a1-7ad2039fbe8b |
| name                        | l1                                   |
| operating_status            | OFFLINE                              |
| project_id                  | e1fe759747844dd1871092efe9079a26     |
| protocol                    | HTTP                                 |
| protocol_port               | 91                                   |
| provisioning_status         | PENDING_CREATE                       |
| sni_container_refs          | []                                   |
| timeout_client_data         | 50000                                |
| timeout_member_connect      | 5000                                 |
| timeout_member_data         | 50000                                |
| timeout_tcp_inspect         | 0                                    |
| updated_at                  | None                                 |
| client_ca_tls_container_ref | None                                 |
| client_authentication       | NONE                                 |
| client_crl_container_ref    | None                                 |
| allowed_cidrs               | None                                 |
| tls_ciphers                 | None                                 |
| tls_versions                | None                                 |
| alpn_protocols              | None                                 |
+-----------------------------+--------------------------------------+
```

**Create a pool**

```
stack@sana:~$ openstack loadbalancer pool create --listener l1  --protocol HTTP --lb-algorithm ROUND_ROBIN --name pool1

+----------------------+--------------------------------------+
| Field                | Value                                |
+----------------------+--------------------------------------+
| admin_state_up       | True                                 |
| created_at           | 2021-07-09T12:48:13                  |
| description          |                                      |
| healthmonitor_id     |                                      |
| id                   | b4383b0e-359f-4699-961f-504e26ce1df0 |
| lb_algorithm         | ROUND_ROBIN                          |
| listeners            | 42199b41-9abe-4554-b9bf-cc3fc6a96432 |
| loadbalancers        | ed1c3237-4a70-435d-88a1-7ad2039fbe8b |
| members              |                                      |
| name                 | pool1                                |
| operating_status     | OFFLINE                              |
| project_id           | e1fe759747844dd1871092efe9079a26     |
| protocol             | HTTP                                 |
| provisioning_status  | PENDING_CREATE                       |
| session_persistence  | None                                 |
| updated_at           | None                                 |
| tls_container_ref    | None                                 |
| ca_tls_container_ref | None                                 |
| crl_container_ref    | None                                 |
| tls_enabled          | False                                |
| tls_ciphers          | None                                 |
| tls_versions         | None                                 |
+----------------------+--------------------------------------+

```
**Create a Member**

```
stack@sana:~$ openstack loadbalancer member create --subnet-id vip_subnet --protocol-port 83 --address 10.0.0.29 pool1 --name m1

+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| address             | 10.0.0.29                            |
| admin_state_up      | True                                 |
| created_at          | 2021-07-09T12:48:37                  |
| id                  | 3d2c90a7-1538-4177-ba82-18f36cd25130 |
| name                | m1                                   |
| operating_status    | NO_MONITOR                           |
| project_id          | e1fe759747844dd1871092efe9079a26     |
| protocol_port       | 83                                   |
| provisioning_status | PENDING_CREATE                       |
| subnet_id           | 8330780e-25fd-4ec7-a6dd-fd412ec81e48 |
| updated_at          | None                                 |
| weight              | 1                                    |
| monitor_port        | None                                 |
| monitor_address     | None                                 |
| backup              | False                                |
+---------------------+--------------------------------------+
```

**Create a health monitor**

```
stack@sana:~$ openstack loadbalancer healthmonitor create --delay 10 --timeout 3 --max-retries 3 --type HTTP --name hm1 pool1

+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| project_id          | e1fe759747844dd1871092efe9079a26     |
| name                | hm1                                  |
| admin_state_up      | True                                 |
| pools               | b4383b0e-359f-4699-961f-504e26ce1df0 |
| created_at          | 2021-07-09T12:49:19                  |
| provisioning_status | PENDING_CREATE                       |
| updated_at          | None                                 |
| delay               | 10                                   |
| expected_codes      | 200                                  |
| max_retries         | 3                                    |
| http_method         | GET                                  |
| timeout             | 3                                    |
| max_retries_down    | 3                                    |
| url_path            | /                                    |
| type                | HTTP                                 |
| id                  | e6a7ae1c-e48a-46c9-a33d-8a41d1889e1a |
| operating_status    | OFFLINE                              |
| http_version        | None                                 |
| domain_name         | None                                 |
+---------------------+--------------------------------------+
```

stack@sana:~$

**Create a L7policy**

```
stack@sana:~$ openstack loadbalancer l7policy create --action REJECT l1 --name policy

+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| listener_id         | 42199b41-9abe-4554-b9bf-cc3fc6a96432 |
| description         |                                      |
| admin_state_up      | True                                 |
| rules               |                                      |
| project_id          | e1fe759747844dd1871092efe9079a26     |
| created_at          | 2021-07-09T12:50:14                  |
| provisioning_status | PENDING_CREATE                       |
| updated_at          | None                                 |
| redirect_pool_id    | None                                 |
| redirect_url        | None                                 |
| redirect_prefix     | None                                 |
| action              | REJECT                               |
| position            | 1                                    |
| id                  | a9f140e0-cb3a-47c4-84ac-32628182b1ba |
| operating_status    | OFFLINE                              |
| name                | policy                               |
| redirect_http_code  | None                                 |
+---------------------+--------------------------------------+
```

stack@sana:~$

**Create a L7 Rule :**

```
stack@sana:~$ openstack loadbalancer l7rule create --type FILE_TYPE --compare-type REGEX --value 'abc' policy

+---------------------+--------------------------------------+
| Field               | Value                                |
+---------------------+--------------------------------------+
| created_at          | 2021-07-09T12:50:50                  |
| compare_type        | REGEX                                |
| provisioning_status | PENDING_CREATE                       |
| invert              | False                                |
| admin_state_up      | True                                 |
| updated_at          | None                                 |
| value               | abc                                  |
| key                 | None                                 |
| project_id          | e1fe759747844dd1871092efe9079a26     |
| type                | FILE_TYPE                            |
| id                  | 15c42445-d48b-4344-8887-c94b79f9ec9f |
| operating_status    | OFFLINE                              |
+---------------------+--------------------------------------+

```
stack@sana:~$

**VThunder Configuration:**

```
vThunder(NOLICENSE)#show running-config
!Current configuration: 99 bytes
!Configuration last updated at 13:50:51 IST Fri Jul 9 2021
!Configuration last saved at 13:50:58 IST Fri Jul 9 2021
!64-bit Advanced Core OS (ACOS) version 5.2.1, build 153 (Dec-11-2020,14:16)
!
!
interface management
  ip address dhcp
!
interface ethernet 1
  name DataPort
  enable
  ip address dhcp
!
!
health monitor e6a7ae1c-e48a-46c9-a33d-8a41d1889e1a
  override-port 91
  interval 10 timeout 3
  method http port 91 expect response-code 200 url GET /
!
slb server e1fe7_10_0_0_29 10.0.0.29
  port 83 tcp
!
slb service-group b4383b0e-359f-4699-961f-504e26ce1df0 tcp
  health-check e6a7ae1c-e48a-46c9-a33d-8a41d1889e1a
  member e1fe7_10_0_0_29 83
!
slb virtual-server ed1c3237-4a70-435d-88a1-7ad2039fbe8b 192.168.12.146
  port 91 http
    name 42199b41-9abe-4554-b9bf-cc3fc6a96432
    extended-stats
    aflex a9f140e0-cb3a-47c4-84ac-32628182b1ba
    service-group b4383b0e-359f-4699-961f-504e26ce1df0
!
!
cloud-services meta-data
  enable
  provider openstack
!
end
!Current config commit point for partition 0 is 0 & config mode is classical-mode
vThunder(NOLICENSE)#

vThunder(NOLICENSE)#show aflex a9f140e0-cb3a-47c4-84ac-32628182b1ba
Name:                    a9f140e0-cb3a-47c4-84ac-32628182b1ba
Syntax:                  Check
Virtual port:            Bind
                         ed1c3237-4a70-435d-88a1-7ad2039fbe8b: 91
Statistics:
    Event HTTP_REQUEST         execute 0 times (0 failures, 0 aborts)
Content:
 when HTTP_REQUEST {

        if { ([HTTP::uri] matches_regex "abc") } {

        HTTP::close

        }

        }
vThunder(NOLICENSE)#
```

### CODE REVIEW

File: a10_octavia/common/config_options.py [hunk 1]
```diff
@@ -271,9 +271,18 @@
...
+    cfg.IntOpt('retry_attempts',
+               default=15,
+               help=_('Retry attempts for Database Entry')),
+    cfg.IntOpt('retry_initial_delay',
+               default=1,
+               help=_('Retry initial delay for Database Entry')),
+    cfg.IntOpt('retry_bakcoff',
+               default=1,
+               help=_('Retry Backoff for Database Entry')),
+    cfg.IntOpt('retry_max',
```
Comment (ytsai-a10) [inline a10_octavia/common/config_options.py:283]: can we change to db_retry_max_delay

File: a10_octavia/common/config_options.py [hunk 2]
```diff
@@ -271,9 +271,18 @@
...
-               help=_('Database Timeout for creating a Database entry'))
+    cfg.IntOpt('retry_attempts',
+               default=15,
+               help=_('Retry attempts for Database Entry')),
+    cfg.IntOpt('retry_initial_delay',
+               default=1,
+               help=_('Retry initial delay for Database Entry')),
+    cfg.IntOpt('retry_bakcoff',
+               default=1,
+               help=_('Retry Backoff for Database Entry')),
```
Comment (ytsai-a10) [inline a10_octavia/common/config_options.py:282]: Can we change to "Retry delay backoff for database fetching"

File: a10_octavia/common/config_options.py [hunk 3]
```diff
@@ -271,9 +271,18 @@
                default=900,
                help=_('Timeout for waiting when vThunder instance is busy. '
                       '(0 for no timeout')),
-    cfg.IntOpt('max_db_timeout',
-               default=60,
-               help=_('Database Timeout for creating a Database entry'))
+    cfg.IntOpt('retry_attempts',
+               default=15,
+               help=_('Retry attempts for Database Entry')),
```
Comment (ytsai-a10) [inline a10_octavia/common/config_options.py:276]: Can we change to  "Retry attempts for database fetching"

File: a10_octavia/common/config_options.py [hunk 4]
```diff
@@ -271,9 +271,18 @@
...
+    cfg.IntOpt('retry_initial_delay',
+               default=1,
+               help=_('Retry initial delay for Database Entry')),
```
Comment (ytsai-a10) [inline a10_octavia/common/config_options.py:279]: Can we change to "Retry initial delay for database fetching"

File: a10_octavia/common/config_options.py [hunk 5]
```diff
@@ -271,9 +271,18 @@
...
+    cfg.IntOpt('retry_bakcoff',
+               default=1,
+               help=_('Retry Backoff for Database Entry')),
+    cfg.IntOpt('retry_max',
+               default=5,
+               help=_('Maximum Retries for Database Entry'))
```
Comment (ytsai-a10) [inline a10_octavia/common/config_options.py:285]: can we change to "Maximum retry delay for database fetching"

**Review by ytsai-a10 — CHANGES_REQUESTED** [review]

File: a10_octavia/common/config_options.py [hunk 6]
```diff
@@ -271,9 +271,18 @@
                default=900,
                help=_('Timeout for waiting when vThunder instance is busy. '
                       '(0 for no timeout')),
-    cfg.IntOpt('max_db_timeout',
-               default=60,
-               help=_('Database Timeout for creating a Database entry'))
+    cfg.IntOpt('retry_attempts',
```
[the remark this answers is not in the document]
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

File: a10_octavia/common/config_options.py [hunk 7]
```diff
@@ -271,9 +271,18 @@
...
+               default=15,
+               help=_('Retry attempts for Database Entry')),
+    cfg.IntOpt('retry_initial_delay',
```
[the remark this answers is not in the document]
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

File: a10_octavia/common/config_options.py [hunk 8]
```diff
@@ -271,9 +271,18 @@
...
+               default=1,
+               help=_('Retry initial delay for Database Entry')),
+    cfg.IntOpt('retry_bakcoff',
```
[the remark this answers is not in the document]
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

File: a10_octavia/common/config_options.py (hunk 1 above)
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

File: a10_octavia/common/config_options.py (hunk 3 above)
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

File: a10_octavia/common/config_options.py (hunk 4 above)
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

File: a10_octavia/common/config_options.py (hunk 5 above)
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

File: a10_octavia/common/config_options.py (hunk 2 above)
Comment (sana-a10) [reply a10_octavia/common/config_options.py]: Fixed

**Review by ytsai-a10 — APPROVED** [review]

**Pull request merged**